internal/component/prometheus/remote/queue/serialization/serializer.go

```diff
@@ -0,0 +1,223 @@
+		if !ok {
+			return actor.WorkerEnd
+		}
+		s.maxItemsBeforeFlush = int(cfg.MaxSignalsInBatch)
+		s.flushFrequency = cfg.FlushFrequency
+		return actor.WorkerContinue
+	default:
+	}
+
+	select {
+	case <-ctx.Done():
+		return actor.WorkerEnd
+	case item, ok := <-s.inbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		err := s.Append(ctx, item)
+		if err != nil {
+			level.Error(s.logger).Log("msg", "unable to append to serializer", "err", err)
+		}
+		return actor.WorkerContinue
+	case item, ok := <-s.metaInbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		err := s.AppendMetadata(ctx, item)
+		if err != nil {
+			level.Error(s.logger).Log("msg", "unable to append metadata to serializer", "err", err)
+		}
+		return actor.WorkerContinue
+	case <-s.flushTestTimer.C:
+		if time.Since(s.lastFlush) > s.flushFrequency {
+			err := s.store(ctx)
+			if err != nil {
+				level.Error(s.logger).Log("msg", "unable to store data", "err", err)
+			}
+		}
+		return actor.WorkerContinue
+	}
+}
+
+func (s *serializer) AppendMetadata(ctx actor.Context, data *types.TimeSeriesBinary) error {
+	s.meta = append(s.meta, data)
+	// If we would go over the max size then send, or if we have hit the flush duration then send.
+	if len(s.meta)+len(s.series) >= s.maxItemsBeforeFlush {
+		return s.store(ctx)
+	} else if time.Since(s.lastFlush) > s.flushFrequency {
+		return s.store(ctx)
+	}
+	return nil
+}
+
+func (s *serializer) Append(ctx actor.Context, data *types.TimeSeriesBinary) error {
+	s.series = append(s.series, data)
+	// If we would go over the max size then send, or if we have hit the flush duration then send.
+	if len(s.meta)+len(s.series) >= s.maxItemsBeforeFlush {
+		return s.store(ctx)
+	} else if time.Since(s.lastFlush) > s.flushFrequency {
+		return s.store(ctx)
+	}
+	return nil
+}
+
+func (s *serializer) store(ctx actor.Context) error {
+	var err error
+	defer func() {
+		s.lastFlush = time.Now()
+	}()
+	// Do nothing if there is nothing.
+	if len(s.series) == 0 && len(s.meta) == 0 {
+		return nil
+	}
+	group := &types.SeriesGroup{
+		Series:   make([]*types.TimeSeriesBinary, len(s.series)),
+		Metadata: make([]*types.TimeSeriesBinary, len(s.meta)),
+	}
+	defer func() {
+		s.storeStats(err)
+		// Return series to the pool, this is key to reducing allocs.
+		types.PutTimeSeriesBinarySlice(s.series)
+		types.PutTimeSeriesBinarySlice(s.meta)
+		s.series = s.series[:0]
+		s.meta = s.series[:0]
+	}()
+
+	// This maps strings to index position in a slice. This is doing to reduce the file size of the data.
+	strMapToInt := make(map[string]uint32)
+	for i, ts := range s.series {
+		ts.FillLabelMapping(strMapToInt)
+		group.Series[i] = ts
+	}
+	for i, ts := range s.meta {
+		ts.FillLabelMapping(strMapToInt)
+		group.Metadata[i] = ts
+	}
+
+	stringsSlice := make([]string, len(strMapToInt))
+	for stringValue, index := range strMapToInt {
+		stringsSlice[index] = stringValue
+	}
+	group.Strings = stringsSlice
```

Comment:
```suggestion
	dictionary := make([]string, len(strMapToInt))
	for stringValue, index := range strMapToInt {
		dictionary[index] = stringValue
	}
	group.dictionary = dictionary
```

I like to use the concept of dictionary here, or lookup table... it makes it easier to figure out what's going on.

Reply:
Do you mean to use an actual map? Or a rename like above?